File: netflix_data_analysis.py
from pyspark.sql import SparkSession
from collections import defaultdict
import sys
import pandas as pd

m_movies_b = None
def movie_ref_mapper_func(value):
    return (value[0], {'movie_id': value[0],
                   'publication': value[1],
                   'movie_name': value[2]})

def parse_csv_mapper_func(value):
    movie_id = ''
    out = []
    for index, line in enumerate(value.value.split('\n')):
        if index == 0:
            movie_id = line.strip(':')
        else:
            value_splits = line.split(',')
            if len(value_splits)==3:
                out.append({'movie_id': movie_id,
                            'user_id': value_splits[0],
                            'rating': value_splits[1],
                            'date': value_splits[2]} )
    return out

def movies_mapper_func(value):
    return (value['movie_id'], value)

def user_mapper_func(value):
    return (value['user_id'], value)

def movies_reducer_func(value):
    cnt = 0
    rating_total = 0
    movie_name = ''
    movie_publication = 0

    for val in value[1][1]:
        movie_name = val['movie_name']
        movie_publication = int(val['publication']) if val['publication'] != 'NULL' else 0

    for val in value[1][0]:
        cnt += 1
        rating_total += int(val['rating'])

    avg = rating_total / cnt
    # return (v[0], movie_name, movie_publication, cnt, avg)
    return ({'movie_id': value[0],
             'movie_name': movie_name,
             'movie_publication': movie_publication,
             'cnt': cnt,
             'avg': avg})

def user_reducer_func(value):
    global m_movies_b
    m_movies = m_movies_b.value

    movies_lst = []
    ratings_dict = defaultdict(list)
    rating_sum = 0
    intersect_count = 0

    for val in value[1]:
        rating = int(val['rating'])
        ratings_dict[rating].append({'movie_id': val['movie_id'],
                                     'rating-dt': val['date']})
        if val['movie_id'] in m_movies:
            intersect_count += 1
            rating_sum += rating

    max_rating = max(ratings_dict.keys())
    highest_rated_films = ratings_dict[max_rating]

    return ({'user_id': value[0],
             'highest_rated_film': highest_rated_films,
             'intersect_count': intersect_count,
             'm_movies_avg_rating': float(rating_sum) / intersect_count if intersect_count > 0 else 0,
             'highest_rating': max_rating,
             'total': len(value[1])})

def get_top_m_movies(M, R, lst):
    lst = sorted(lst, key=lambda x: (-x['cnt'], -int(x['movie_publication']), x['movie_name']))
    return set([x['movie_id'] for x in lst if x['cnt'] >= R][:M])

def get_top_u_users(M, U, lst, movies_dict):
    lst = [v for v in lst if v['intersect_count'] == M]
    lst = sorted(lst, key=lambda x: (float(x['m_movies_avg_rating']), int(x['user_id'])))

    out = []
    for i, v in enumerate(lst):
        if i == U:
            break

        movie_lst = []
        for movie in v['highest_rated_film']:
            movie_lst.append({'rating_dt': movie['rating-dt'],
                              'publication': movies_dict[movie['movie_id']]['publication'],
                              'movie_name': movies_dict[movie['movie_id']]['movie_name']
                              })
        movie = sorted(movie_lst, key=lambda x: (-int(x['publication']), x['movie_name']))[0]
        out.append({'user_id': v['user_id'],
                    'year_release': movie['publication'],
                    'rating_dt': movie['rating_dt'],
                    'movie': movie['movie_name'],
                    'm_movies_avg_rating': v['m_movies_avg_rating'],
                    'highest_rating': v['highest_rating']})
    return out

def get_movies_dict(movies_ref_rdd):
    movies_df = movies_ref_rdd.toDF().toPandas()
    movies_dict = {}
    for indx, row in movies_df.iterrows():
        movies_dict[row[0]] = {'publication': row[1],
                               'movie_name': row[2]}
    return movies_dict

def main():
    args = sys.argv[1:]
    spark = SparkSession.builder.appName("netflix_data_analysis").getOrCreate()
    movies_rdd = spark.read.text(args[3], wholetext=True).rdd
    movies_ref_rdd = spark.read.csv(args[4]).rdd
    movie_ref_mapper_rdd = movies_ref_rdd.map(movie_ref_mapper_func).groupByKey()
    movies_dict = get_movies_dict(movies_ref_rdd)
    parse_csv_mapper_rdd = movies_rdd.flatMap(parse_csv_mapper_func).cache()
    movies_mapper_rdd = parse_csv_mapper_rdd.map(movies_mapper_func).groupByKey()
    movies_reducer_rdd = movies_mapper_rdd.join(movie_ref_mapper_rdd).map(movies_reducer_func)
    movies_list = spark.createDataFrame(movies_reducer_rdd).toPandas().to_dict('records')
    M_conf = int(args[0])
    R_conf = int(args[1])
    U_conf = int(args[2])

    global m_movies_b
    m_movies = get_top_m_movies(M_conf, R_conf, movies_list)
    m_movies_b = spark.sparkContext.broadcast(m_movies)
    user_mapper_rdd = parse_csv_mapper_rdd.map(user_mapper_func).groupByKey()
    user_reducer_rdd = user_mapper_rdd.map(user_reducer_func)
    users_list = spark.createDataFrame(user_reducer_rdd).toPandas().to_dict('records')
    top_u_users_lst = get_top_u_users(M_conf, U_conf, users_list, movies_dict)
    df = pd.DataFrame(top_u_users_lst)
    df.to_csv(args[5], index=False, sep= '\t', columns=['user_id', 'movie', 'year_release', 'rating_dt'] )


if __name__ == '__main__':
    main()

File: netflix_data_analysis.scala
package org.example.helloworld

import scala.App
import org.apache.spark
import org.apache.spark.SparkContext
import org.apache.spark.sql.SQLContext
import org.apache.spark.sql.functions.col

import scala.collection.mutable
import scala.util.control.Breaks

class Movie(val movieId:String, val publication:Int, val movieName:String) extends Serializable
class UserRating(val movieId: String, val userId: String, val rating: Int, val ratingDate: String) extends Serializable
class MovieStat(val movieId: String, val movieName: String, val publication: Int, val ratingsCount: Long, val averageRating: Float) extends Serializable
class UserStat(val userId: String, val highestRatedFilms: Array[UserRating], val intersectCount: Int, val mMoviesAvgRating: Float) extends Serializable
class Result(val userId: String, val highestRatedFilm: String, val mMoviesAvgRating: Float, val yearRelease:Int, val ratingDt: String )

object UtilityFunctions extends Serializable {
  def getMoviesDict(value: Array[org.apache.spark.sql.Row] ): Map[String,Movie] =   {
    var obj: Map[String, Movie] = Map()
    var lineSplits: Array[String] = Array()
    for(i <- value.indices){
      lineSplits = value(i)(0).toString.split(",")
      obj+= (lineSplits(0).trim -> new Movie(lineSplits(0).trim, if (lineSplits(1)!="NULL") lineSplits(1).trim.toInt else 0, lineSplits(2).trim))
    }
    obj
  }

  def getTopMoviesList(value: Array[org.apache.spark.sql.Row]): Set[String] = {
    var moviesList:Set[String] = Set()
    for (i<-value.indices){
      moviesList+=value(i)(0).toString
    }
    moviesList
  }

  def printResult(result: Array[Result]):Unit ={
    for (i<-result.indices){
      println(result(i).userId, result(i).highestRatedFilm, result(i).yearRelease, result(i).ratingDt)
    }
  }

  def getTopMUsers(M:Int, U:Int, userStatsArray: Array[UserStat], moviesDict: Map[String,Movie]): Array[Result] = {

    val topUsers = userStatsArray.sortBy(r => (r.mMoviesAvgRating, r.userId.toInt))
    var results: Array[Result] = Array()
    var userRatings: Array[(String, String, Int)] = Array()
    var ratingsArray: Array[UserRating] = Array()

    val loop = new Breaks
    loop.breakable{
      for (i <- topUsers.indices) {
        if (results.length == U) {
          loop.break
        }
        userRatings = Array()
        ratingsArray = topUsers(i).highestRatedFilms
        for (j <- ratingsArray.indices) {
          val rating = ratingsArray(j)
          userRatings :+= (rating.ratingDate, moviesDict(rating.movieId).movieName, moviesDict(rating.movieId).publication)
        }
        val highestRatedFilm = userRatings.sortBy(r => (-r._3, r._2))
        if (topUsers(i).intersectCount == M) {
          results :+= new Result(topUsers(i).userId, highestRatedFilm(0)._2, topUsers(i).mMoviesAvgRating, highestRatedFilm(0)._3, highestRatedFilm(0)._1)
        }
      }
    }
    results
  }
}

object SimpleApp {
  def main(args: Array[String]): Unit = {
    val conf = new spark.SparkConf().setAppName("scala spark")
    val sc = new SparkContext(conf)
    val sqlContext: SQLContext = new org.apache.spark.sql.SQLContext(sc)
    import sqlContext.implicits._
    val mConfBroadcast = sc.broadcast(5)
    val rConfBroadcast = sc.broadcast(10)
    val UConfBroadcast = sc.broadcast(25)

    val moviesRefRDD = sc.textFile("*")
    val moviesRDD = sc.wholeTextFiles("*")



    val movieRefMapperRDD = moviesRefRDD.map(value => {
      val splits: Array[String] = value.split(',')
      (splits(0), new Movie(splits(0), if (splits(1) != "NULL") splits(1).toInt else 0, splits(2)))
    }).groupByKey()

    val parseCSVMapperRDD = moviesRDD.flatMap(value=>{
      var movieId: String = ""
      val lineSplits: Array[String] = value._2.split('\n')
      var out: Array[UserRating] = Array()
      var valueSplits: Array[String] = Array()

      for (i <- lineSplits.indices) {
        if (i == 0) {
          movieId = lineSplits(i).trim.stripSuffix(":")
        }
        else {
          valueSplits = lineSplits(i).split(",")
          if (valueSplits.length == 3) {
            out :+= new UserRating(movieId, valueSplits(0).trim, if (valueSplits(1)!="NULL") valueSplits(1).trim.toInt else 0, valueSplits(2).trim)
          }
        }
      }
      out
    }).cache()

    val moviesMapperRDD = parseCSVMapperRDD.map(value=>{
      (value.movieId, value)
    }).groupByKey()

    val movieReducerRDD = moviesMapperRDD.join(movieRefMapperRDD).map(value=>{
      val ratingsIter: Iterator[UserRating] = value._2._1.iterator
      val movieIter: Iterator[Movie] = value._2._2.iterator
      var count: Long = 0
      var ratingTotal: Long = 0
      var movieName: String = ""
      var moviePublication: Int = 0

      while (movieIter.hasNext) {
        val movieObj: Movie = movieIter.next()
        movieName = movieObj.movieName
        moviePublication = movieObj.publication
      }

      while (ratingsIter.hasNext) {
        val ratingsObj: UserRating = ratingsIter.next()
        count += 1
        ratingTotal += ratingsObj.rating
      }
      val average: Float = ratingTotal / count
      (value._1, movieName, moviePublication, count, average)
    })

    val dfRDD = movieReducerRDD.toDF("movieId", "movieName", "publication", "ratingsCount", "averageRating")
    val dfRDD1 = dfRDD.sort(col("ratingsCount").desc, col("publication").desc, col("movieName").asc).filter(dfRDD("ratingsCount")>rConfBroadcast.value).take(mConfBroadcast.value)
    val movies_lst = UtilityFunctions.getTopMoviesList(dfRDD1)
    val broadcastMoviesList = sc.broadcast(movies_lst)

    val userMapperRDD = parseCSVMapperRDD.map(value=>{
      (value.userId, value)
    }).groupByKey()

    val userReducerRDD = userMapperRDD.map(value=> {
      val ratingsIter: Iterator[UserRating] = value._2.iterator
      val ratingsDict = new mutable.HashMap[Int, Array[UserRating]]() {
        override def default(key: Int): Array[UserRating] = Array()
      }
      var ratingSum: Long = 0
      var intersectCount: Int = 0

      while (ratingsIter.hasNext) {
        val ratingsObj: UserRating = ratingsIter.next()
        ratingsDict(ratingsObj.rating) :+= ratingsObj
        if (broadcastMoviesList.value.contains(ratingsObj.movieId)) {
          intersectCount += 1
          ratingSum += ratingsObj.rating
        }
      }
      new UserStat(value._1, if (ratingsDict.nonEmpty) ratingsDict(ratingsDict.keys.iterator.max) else Array(), intersectCount, if (intersectCount > 0) {
        ratingSum.toFloat / intersectCount
      } else 0)
    })

    val moviesDict = UtilityFunctions.getMoviesDict(moviesRefRDD.toDF().collect)
    val result = UtilityFunctions.getTopMUsers(mConfBroadcast.value, UConfBroadcast.value, userReducerRDD.collect, moviesDict)
    UtilityFunctions.printResult(result)
  }
}


